ecdsa/party/implementation.go

```diff
@@ -0,0 +1,346 @@
+package party
+
+import (
+	"context"
+	"crypto/ecdsa"
+	"encoding/json"
+	"errors"
+	"fmt"
+	"math/big"
+	"os"
+	"path"
+	"runtime"
+	"sort"
+	"sync"
+	"time"
+
+	"github.com/bnb-chain/tss-lib/v2/common"
+	"github.com/bnb-chain/tss-lib/v2/ecdsa/keygen"
+	"github.com/bnb-chain/tss-lib/v2/ecdsa/signing"
+	"github.com/bnb-chain/tss-lib/v2/tss"
+)
+
+type KeygenHandler struct {
+	LocalParty  tss.Party
+	StoragePath string
+	// communication channels
+	ProtocolEndOutput <-chan *keygen.LocalPartySaveData
+
+	SavedData *keygen.LocalPartySaveData
+}
+
+type SingleSigner struct {
+	// Time represents the time when the first message was received.
+	// Given a timeout parameter, bookkeeping and cleanup will use this parameter.
+	time.Time
+
+	// used as buffer for messages received before starting signing.
+	// will be consumed once signing starts.
+	MessageBuffer []tss.ParsedMessage
+
+	// nil if not started signing yet.
+	// once a request to sign was received (via AsyncRequestNewSignature), this will be set,
+	// and used.
+	LocalParty tss.Party
+	sync.Once
+}
+
+func (s *SingleSigner) sortMessageBuffer() {
+	sort.Slice(s.MessageBuffer, func(i, j int) bool {
+		return findRound(s.MessageBuffer[i]) < findRound(s.MessageBuffer[j])
+	})
+}
+
+type SigningHandler struct {
+	Mtx sync.Mutex
+
+	DigestToSigner map[string]*SingleSigner
+
+	SigPartReadyChan chan *common.SignatureData
+}
+
+func (s *SigningHandler) getSignerOrCacheMessage(message tss.ParsedMessage) (*SingleSigner, *tss.Error) {
+	s.Mtx.Lock()
+
+	signer, ok := s.DigestToSigner[string(message.WireMsg().GetDigest())]
+	if !ok {
+
+		s.DigestToSigner[string(message.WireMsg().GetDigest())] = &SingleSigner{
+			Time:          time.Now(),
+			LocalParty:    nil,
+			MessageBuffer: []tss.ParsedMessage{message},
+		}
+
+		s.Mtx.Unlock()
+
+		return nil, nil
+	}
+
+	// haven't been requested to sign this digest yet.
+	if signer.LocalParty == nil {
+		s.DigestToSigner[string(message.WireMsg().GetDigest())].MessageBuffer = append(signer.MessageBuffer, message)
+
+		s.Mtx.Unlock()
+
+		return nil, nil
+	}
+
+	s.Mtx.Unlock()
+
+	var e *tss.Error
```

Comment:
Since the localParty can be initialised by AsyncSign, we need to ensure that, in any case of these functions that return a valid single signer, it is fully operational.
Maybe we should chat about this?